docs/testing.md

```diff
@@ -0,0 +1,204 @@
+
+> Please note that the terminology around what integration testing means is not always consistent. Also, the line between what is a unit test and what is an integration test may not always be clear. As a rule of thumb for this document, when your test
+>
+> - Uses an external system (such as the thermometer in the mocking paragraph)
+> - Does a network call to other application
+> - Does any kind of file or database I/O
+>
+> Then it falls into the “integration test” category.
+
+## Component Tests
+
+React components are responsible for rendering your app, and users will directly interact with their output. Even if your app's business logic has high testing coverage and is correct, without component tests you may still deliver a broken UI to your users. Component tests could fall into both unit and integration testing, and since they are such a core part of React Native, we'll cover them in separate paragraphs.
+
+For testing React components, there are two things you may want to test:
+
+- Interaction: to ensure the component behaves correctly when interacted with by a user (eg. when user presses a button)
+- Rendering: to ensure the component render output used by React is correct
+
+For example, if you have a button that has an `onPress` listener, you want to test that the button both appears correctly and that tapping the button is correctly handled by the component.
+
+There are several libraries that can help you testing these:
+
+- React’s [test renderer](https://reactjs.org/docs/test-renderer.html) is a small library provided by Facebook useful for snapshot testing. It “renders” your component view hierarchy to pure JavaScript objects held in memory. You can then make assertions on those objects in your Jest tests.
+- [`react-native-testing-library`](https://github.com/callstack/react-native-testing-library) builds on top of React’s test renderer and adds `fireEvent` and `query` apis described in the next paragraph.
+- [`@testing-library/react-native`](https://www.native-testing-library.com/) is another alternative that also builds on top of React’s test renderer and adds `fireEvent` and `query` apis described in the next paragraph.
+
+> Note that component tests are only JavaScript tests running on Node, they do _not_ take into account any iOS or Android code which is backing the React Native components. It follows that they cannot give you a 100% confidence that everything works for the user - if there is a bug in the iOS or Android code, they will not find it.
+
+### Testing User Interactions
+
+Aside from rendering some UI, your components handle events like `onChangeText` for `TextInput` or `onPress` for `Button`. They may also contain other functions and event callbacks. Consider the following example:
+
+```js
+class GroceryShoppingListCreator extends React.Component {
```

Comment:
I'll see what I can do, I want this

> you could write your tests to call `addNewItemToShoppingList` directly, and while such test works, it's very close to the implementation details of your component and would break if you refactor the component.

to make sense and if I add hooks, things will become a little less straightforward imho

Reply:
@vonovak We are using hooks and function components in the docs exclusively going forward. If you can do this, I can merge!

Reply:
I know it will be a little less obvious, but it is/will be the more widespread usecase, meaning investing in conveying this right with hooks will pay off more than sticking with the easier explanation in classes.

Reply:
@rachelnabors I rewrote the example with hooks, and I think it's for the better.